reactioncommerce/api-plugin-payments-stripe-sca
# WIP: Stripe Payments Intents API SCA-compliant Payment Plugin

Signed-off-by: Janus Reith <mail@janusreith.de>

Resolves #1 and [#5304](https://github.com/reactioncommerce/reaction/issues/5304)(Reaction Repo)
Impact: **breaking|major**
Type: **feature|refactor**

## Issue
The existing [Stripe Payment Plugin](https://github.com/reactioncommerce/api-plugin-payments-stripe) relies on a deprecated API, and is not usable in the EU anymore (potentially similar regulations outside EU too) due to [Strong Customer Authentication](https://stripe.com/docs/strong-customer-authentication).

## Solution
This PR adds a new Payment Plugin for the [Stripe Payment Intents API](https://stripe.com/docs/payments/payment-intents/quickstart), which implements necessary requirements for SCA.
It covers creating a Payment Intent using a new GraphQL mutation `createStripePaymentIntent` and the four default Payment provider actions: Create a payment, Capture a payment, List refund, Create a refund.

Besides the stripe credit cart from, this API plugin also works with the wide range of other Payment solutions enabled through stripe, like Google Pay or Apple Pay, since these also rely on the same Payment Intents API.

## Breaking changes
This plugin should be able to run independently from the existing stripe plugin and should, although not verified yet, be able to run side by side.
The storefront might receive a breaking change if we replace the Stripe integration there with the SCA-compliant one, will elaborate further in a separate issue.


## Testing
1. Add the new Stripe integration to your storefront(Instructions WIP, I have it prepared but figuring out wether to make it a separate package and best practices for distribution in general. Could also PR it right to storefront for now. )
2. Add the Stripe API Keys to both storefront and API, Keys are the same format as before.
2. Create a cart and proceed though the checkout steps
3. Complete the order by using [Stripe Testing CC numbers](https://stripe.com/docs/testing). Be sure to check with 3D-secure enabled cards too see 2FA in action.
4. Capture the payment for the new order in the backend
5. Issue a partial refund, see that the created refund(s) are listed.

Maybe also check back with your Stripe Dashboard to confirm transactions.

## Todo
Integration tests for the four payment plugin functions and the token mutation need to be added.

## Further thinking
To keep things simple, `createStripePaymentIntent` will for now take the `cartId` and optionally the `cartToken`(similar to the anonymous/account cart when adding cart items) and create a payment intent in stripe for that cart total.
This reflects the way existing Payment plugins works on the cart level.
Other use-cases might be direct single-product checkouts which skip a cart entirely. (Did this for client projects, works).
We could make providing a `cartId` optional and optionally accept `productConfiguration`(`productId` + `variantId`) and the id of a fulfillment Method instead to calculate a total out of that.
